src/api.ts

```diff
@@ -32,214 +52,199 @@ const fleekBucket = env.get("STORAGE_BUCKET").asString();
+      const userAccount = JSON.parse(fileContents) as UserAccount;
+      userBids = userAccount.bids;
+    } catch {
+      // intentional
     }
 
-    //check signature recovers correct account
-    const bidMessage = await encodeBid(
-      dto.auctionId,
-      dto.bidAmount,
-      dto.contractAddress,
-      dto.tokenId,
-      dto.minimumBid,
-      dto.startBlock,
-      dto.expireBlock
-    );
-
-    const unsignedMessage = await zAuctionContract.toEthSignedMessageHash(
-      bidMessage
-    );
-    const recoveredAccount = await zAuctionContract.recover(
-      unsignedMessage,
-      dto.signedMessage
-    );
-
-    if (recoveredAccount != dto.account) {
-      return res.status(405).send({
-        message:
-          "Account sent and account recovered from signature do not match",
-      });
+    return res.status(200).send(userBids);
+  }
+);
+
+// Creates a new bid for an auction once signed by user
+router.post(
+  "/bids",
+  limiter,
+  async (
+    req: express.Request,
+    res: express.Response,
+    next: express.NextFunction
+  ) => {
+    if (!validateBidPostSchema(req.body)) {
+      return res.status(400).send(validateBidPostSchema.errors);
     }
 
-    // try to pull auction from fleek with given auctionId
-    const dateNow = new Date();
-
-    let auction: Maybe<Auction>;
+    const dto: BidPostDto = req.body as BidPostDto;
 
-    const auctionFileKey = nftId;
-    const auctionFile = await storage.safeDownloadFile(auctionFileKey);
+    try {
+      // Instantiate contracts
+      const erc20Contract = await getTokenContract();
+      const zAuctionContract: Zauction = await getZAuctionContract();
 
-    if (auctionFile.exists) {
-      auction = JSON.parse(auctionFile.data) as Auction;
-    } else {
-      auction = {
+      const params: BidParams = {
+        account: dto.account,
+        auctionId: dto.auctionId,
+        bidAmount: dto.bidAmount,
+        contractAddress: dto.contractAddress,
+        tokenId: dto.tokenId,
+        minimumBid: dto.minimumBid,
+        startBlock: dto.startBlock,
+        expireBlock: dto.expireBlock
+      }
+
+      // Perform necessary checks to ensure account is able to make the bid
+      // Check balance, block number, if consumed, if account recoverable
+      const verification: VerifyBidResponse = await verifyEncodedBid(
+        params,
+        dto.signedMessage,
+        erc20Contract,
+        zAuctionContract
+      );
+
+      if (!verification.pass) {
+        return res
+          .status(verification.status)
+          .send({ message: verification.message });
+      }
+
+      // Try to pull auction from fleek with given auctionId
+      const nftId = calculateNftId(dto.contractAddress, dto.tokenId);
+      const auctionFileKey = nftId;
+      const auctionFile = await storage.safeDownloadFile(auctionFileKey);
+      const dateNow = new Date();
+      
+      const newBid: Bid = {
+        account: dto.account,
```

Comment:
minor nit/hack/simplification you could do here is doing
```
{
  ...params,
  date: dateNow.getTime(),
  tokenId: dto.tokenId,
  contractAddress: dto.contractAddress,
}
```

Reply:
Ahh nice I love that, the only item `params` doesn't have is date so this is a nice simplification